File: roblox_plugin/src/Server/formatTypes/dict.ts
//!native
import * as ProcessFormats from "../processFormats";
import SendMessagesThread from "../sendMessages";
import ReceiveMessagesThread from "../receiveMessages";

interface DictFormatData {
	type: "dict";
	index: ProcessFormats.FormatData;
	value: ProcessFormats.FormatData;
}
const DATA_SIZE = 4;
export function read(
	receiveThread: ReceiveMessagesThread,
	args: defined[],
	formatData: DictFormatData,
	masks: Map<string, boolean>,
) {
	const [readBuffer, readOffset] = ReceiveMessagesThread.readBuffer(receiveThread, DATA_SIZE);
	const dictSize = buffer.readu32(readBuffer, readOffset);
	const dict = new Map<defined, defined>();

	for (const _ of $range(1, dictSize)) {
		const k: defined[] = [];
		const v: defined[] = [];
		ReceiveMessagesThread.parse(receiveThread, k, formatData.index, masks);
		ReceiveMessagesThread.parse(receiveThread, v, formatData.value, masks);
		dict.set(k[0], v.size() > 1 ? v : v[0]);
	}
	args.push(dict);
}
export function write(
	sendThread: SendMessagesThread,
	args: defined[],
	argsCount: number,
	formatData: DictFormatData,
	masks: Map<string, boolean>,
): number {
	const dict = args[argsCount] as Map<defined, defined>;
	const dictSize = dict.size();
	const data = buffer.create(DATA_SIZE);
	buffer.writeu32(data, 0, dictSize);
	SendMessagesThread.writeBuffer(sendThread, data, DATA_SIZE);

	if (dictSize > 0) {
		const [_, testValue] = next(dict);
		if (typeIs(testValue, "table")) {
			for (const [k, v] of dict) {
				SendMessagesThread.parse(sendThread, [k], 0, formatData.index, masks);
				SendMessagesThread.parse(sendThread, v as defined[], 0, formatData.value, masks);
			}
		} else {
			for (const [k, v] of dict) {
				SendMessagesThread.parse(sendThread, [k], 0, formatData.index, masks);
				SendMessagesThread.parse(sendThread, [v], 0, formatData.value, masks);
			}
		}
	}
	return argsCount + 1;
}

File: blender_roblox_sync/server/format_types/dict.py
import struct

format = struct.Struct("<L")
format_size = format.size
unpack_from = format.unpack_from
pack = format.pack

def read(self, format_data, **masks):
	dict_len, = unpack_from(*self.read_buffer(format_size))
	dict_value = {}
	for _ in range(dict_len):
		k, = self.parse(format_data["index"], **masks)
		v = self.parse(format_data["value"], **masks)
		dict_value[k] = v if len(v) > 1 else v[0]
	return (dict_value,)

def write(self, args, args_count, format_data, **masks):
	dict_items = args[args_count].items()
	dict_len = len(dict_items)
	self.write_buffer(pack(dict_len), format_size)
	if dict_len > 0:
		test_value = next(iter(dict_items))[1]
		values_type = type(test_value)
		if (values_type is list) or (values_type is tuple):
			for k, v in dict_items:
				self.parse((k,), 0, format_data["index"], **masks)
				self.parse(v, 0, format_data["value"], **masks)
		else:
			for k, v in dict_items:
				self.parse((k,), 0, format_data["index"], **masks)
				self.parse((v,), 0, format_data["value"], **masks)
	return args_count + 1